src/v/cloud_topics/reader/placeholder_extent_reader.cc

```diff
@@ -0,0 +1,196 @@
+      model::record_batch_reader rdr,
+      const storage::log_reader_config& cfg,
+      cloud_storage_clients::bucket_name bucket,
+      cloud_io::remote_api<>& api,
+      cloud_io::basic_cache_service_api<>& cache,
+      retry_chain_node* rtc)
+      : _underlying(std::move(rdr))
+      , _config(cfg)
+      , _bucket(std::move(bucket))
+      , _api(api)
+      , _cache(cache)
+      , _rtc(rtc) {}
+
+    bool is_end_of_stream() const override {
+        bool is_eos = _config.start_offset >= _config.max_offset
+                      || _underlying.is_end_of_stream();
+        return is_eos;
+    }
+
+    ss::future<model::record_batch_reader::storage_t>
+    do_load_slice(model::timeout_clock::time_point tm) override {
+        struct consumer {
+            consumer(
+              storage::log_reader_config* config,
+              ss::circular_buffer<model::record_batch>* rb)
+              : _config(config)
+              , _read_buffer(rb) {}
+
+            ss::future<ss::stop_iteration> operator()(model::record_batch rb) {
+                if (rb.header().base_offset < _config->start_offset) {
+                    // Skip data which was already consumed
+                    co_return ss::stop_iteration::no;
+                }
+                if (rb.header().last_offset() > _config->max_offset) {
+                    // Stop on EOS
+                    co_return ss::stop_iteration::yes;
+                }
+                // Account bytes based on placeholder content, not header
+                _num_bytes += get_extent_size(rb);
+                auto last_offset = rb.header().last_offset();
+                _read_buffer->push_back(std::move(rb));
+                _config->start_offset = model::next_offset(last_offset);
+                bool done = _config->start_offset >= _config->max_offset
+                            || _num_bytes > _config->max_bytes;
+                auto res = done ? ss::stop_iteration::yes
+                                : ss::stop_iteration::no;
+                co_return res;
+            }
+
+            void end_of_stream() {}
+
+        private:
+            storage::log_reader_config* _config;
+            ss::circular_buffer<model::record_batch>* _read_buffer;
+            size_t _num_bytes{0};
+        };
+
+        ss::circular_buffer<model::record_batch> read_buffer;
+        consumer cons(&_config, &read_buffer);
+        co_await _underlying.consume(cons, tm);
+
+        auto extents = co_await materialize_sorted_run(
+          std::move(read_buffer), _bucket, &_api, &_cache, &_rtc);
+        if (extents.has_error()) {
+            vlog(
+              cd_log.error,
+              "Failed to materialize sorted run: {}",
+              extents.error().message());
+            throw std::system_error(extents.error());
+        }
+
+        ss::circular_buffer<model::record_batch> slice;
+        for (auto& e : extents.value()) {
+            slice.push_back(e.make_raft_data_batch());
```

Comment:
it isn't obvious why we are converting placeholder batches to raft batches. i'm guessing that it has something to do with the rest of the system expecting raft data batches? are there parts of the raft batch that can't be filled in completely, or is it identical, etc...?

Reply:
The placeholder batch doesn't have the data, so we're bringing the data and replacing placeholder batch with the raft_data batch. We need to make this a raft-data batch because the Kafka layer will filter out anything which isn't a raft-data batch. We can introduce another batch type (e.g. materialized_placeholder) and teach Kafka layer to use it but IMO this will only add complexity.

Reply:
> We need to make this a raft-data batch because the Kafka layer will filter out anything which isn't a raft-data batch

this is what i was looking for. thanks